File: Problema2/Problema2/solucion.py
import time
import sys

def generar_coeficientes_pascal(n):

    if n == 0:
        return [1]
    
    # Inicializar con la primera fila del triángulo
    coeficientes = [1]
    
    for i in range(1, n + 1):
        nuevo_coeficiente = [1]  # Primer coeficiente siempre es 1
        
        # Calcular coeficientes intermedios
        for j in range(1, i):
            nuevo_coeficiente.append(coeficientes[j - 1] + coeficientes[j])
        
        nuevo_coeficiente.append(1)  # Último coeficiente siempre es 1
        coeficientes = nuevo_coeficiente
    
    return coeficientes

def mostrar_polinomio(coeficientes):

    n = len(coeficientes) - 1
    terminos = []
    
    for i, coef in enumerate(coeficientes):
        exponente = n - i
        
        if coef == 0:
            continue
            
        if exponente == 0:
            terminos.append(f"{coef}")
        elif exponente == 1:
            if coef == 1:
                terminos.append("x")
            else:
                terminos.append(f"{coef}x")
        else:
            if coef == 1:
                terminos.append(f"x^{exponente}")
            else:
                terminos.append(f"{coef}x^{exponente}")
    
    return " + ".join(terminos)

def calcular_polinomio(coeficientes, x):

    n = len(coeficientes) - 1
    resultado = 0
    pasos = []
    
    print(f"\nCálculo paso a paso para x = {x}:")
    print(f"Polinomio: {mostrar_polinomio(coeficientes)}")
    print("-" * 50)
    
    for i, coef in enumerate(coeficientes):
        exponente = n - i
        termino = coef * (x ** exponente)
        resultado += termino
        
        if exponente == 0:
            pasos.append(f"{coef} × {x}^{exponente} = {coef} × 1 = {termino}")
        else:
            pasos.append(f"{coef} × {x}^{exponente} = {coef} × {x**exponente} = {termino}")
    
    # Mostrar todos los pasos
    for i, paso in enumerate(pasos):
        print(f"Término {i+1}: {paso}")
    
    print("-" * 50)
    print(f"Resultado final: {resultado}")
    print(f"Verificación: ({x} + 1)^{n} = {resultado}")
    
    return resultado

def main():
    try:
        # Leer n del usuario
        n = int(input("Ingrese el valor de n (entero no negativo): "))
        if n < 0:
            print("Error: n debe ser un número entero no negativo")
            return
        
        # Medir tiempo de generación de coeficientes
        inicio_generacion = time.time()
        coeficientes = generar_coeficientes_pascal(n)
        tiempo_generacion = time.time() - inicio_generacion
        
        print(f"\n=== RESULTADOS PARA n = {n} ===")
        print(f"Coeficientes: {coeficientes}")
        print(f"Polinomio: {mostrar_polinomio(coeficientes)}")
        print(f"Tiempo de generación: {tiempo_generacion:.6f} segundos")
        
        # Calcular para un valor específico de x
        x = float(input("\nIngrese el valor de x para calcular el polinomio: "))
        
        inicio_calculo = time.time()
        calcular_polinomio(coeficientes, x)
        tiempo_calculo = time.time() - inicio_calculo
        
        print(f"Tiempo de cálculo: {tiempo_calculo:.6f} segundos")
        
        # Guardar resultados en archivo
        with open(f"resultados_polinomio_n{n}.txt", "w", encoding="utf-8") as archivo:
            archivo.write(f"RESULTADOS PARA (x+1)^{n}\n")
            archivo.write("=" * 50 + "\n")
            archivo.write(f"Coeficientes: {coeficientes}\n")
            archivo.write(f"Polinomio: {mostrar_polinomio(coeficientes)}\n")
            archivo.write(f"Tiempo de generación: {tiempo_generacion:.6f} segundos\n")
            archivo.write(f"Valor calculado para x={x}: {calcular_polinomio(coeficientes, x)}\n")
            archivo.write(f"Tiempo de cálculo: {tiempo_calculo:.6f} segundos\n")
        
        print(f"\nResultados guardados en 'resultados_polinomio_n{n}.txt'")
        
        # Prueba específica para n=100
        if n == 100:
            print(f"\n=== PRUEBA ESPECÍFICA PARA n=100 ===")
            print(f"Número de coeficientes: {len(coeficientes)}")
            print(f"Coeficiente máximo: {max(coeficientes)}")
            
    except ValueError:
        print("Error: Por favor ingrese un número válido")
    except Exception as e:
        print(f"Error inesperado: {e}")

if __name__ == "__main__":
    main()

File: Problema2/Problema2/solucion.cpp
#include <iostream>
#include <vector>
#include <string>
#include <cmath>
#include <chrono>
#include <fstream>
#include <iomanip>

using namespace std;
using namespace std::chrono;

// Función para generar coeficientes usando el triángulo de Pascal
vector<long long> generarCoeficientesPascal(int n) {
    vector<long long> coeficientes;
    
    if (n == 0) {
        coeficientes.push_back(1);
        return coeficientes;
    }
    
    // Inicializar con la primera fila
    coeficientes.push_back(1);
    
    for (int i = 1; i <= n; i++) {
        vector<long long> nuevoCoeficiente;
        nuevoCoeficiente.push_back(1); // Primer coeficiente siempre es 1
        
        // Calcular coeficientes intermedios
        for (int j = 1; j < i; j++) {
            nuevoCoeficiente.push_back(coeficientes[j - 1] + coeficientes[j]);
        }
        
        nuevoCoeficiente.push_back(1); // Último coeficiente siempre es 1
        coeficientes = nuevoCoeficiente;
    }
    
    return coeficientes;
}

// Función para mostrar el polinomio en formato legible
string mostrarPolinomio(const vector<long long>& coeficientes) {
    int n = coeficientes.size() - 1;
    string polinomio = "";
    
    for (int i = 0; i < coeficientes.size(); i++) {
        int exponente = n - i;
        long long coef = coeficientes[i];
        
        if (coef == 0) continue;
        
        if (i > 0) polinomio += " + ";
        
        if (exponente == 0) {
            polinomio += to_string(coef);
        } else if (exponente == 1) {
            if (coef == 1) {
                polinomio += "x";
            } else {
                polinomio += to_string(coef) + "x";
            }
        } else {
            if (coef == 1) {
                polinomio += "x^" + to_string(exponente);
            } else {
                polinomio += to_string(coef) + "x^" + to_string(exponente);
            }
        }
    }
    
    return polinomio;
}

// Función para calcular el polinomio mostrando pasos
double calcularPolinomioPasos(const vector<long long>& coeficientes, double x) {
    int n = coeficientes.size() - 1;
    double resultado = 0.0;
    
    cout << "\nCálculo paso a paso para x = " << x << ":" << endl;
    cout << "Polinomio: " << mostrarPolinomio(coeficientes) << endl;
    cout << string(50, '-') << endl;
    
    for (int i = 0; i < coeficientes.size(); i++) {
        int exponente = n - i;
        long long coef = coeficientes[i];
        double termino = coef * pow(x, exponente);
        resultado += termino;
        
        if (exponente == 0) {
            cout << "Término " << (i + 1) << ": " << coef << " × " 
                 << x << "^" << exponente << " = " << coef << " × 1 = " << termino << endl;
        } else {
            cout << "Término " << (i + 1) << ": " << coef << " × " 
                 << x << "^" << exponente << " = " << coef << " × " 
                 << pow(x, exponente) << " = " << termino << endl;
        }
    }
    
    cout << string(50, '-') << endl;
    cout << "Resultado final: " << resultado << endl;
    cout << "Verificación: (" << x << " + 1)^" << n << " = " << resultado << endl;
    
    return resultado;
}

int main() {
    try {
        int n;
        cout << "Ingrese el valor de n (entero no negativo): ";
        cin >> n;
        
        if (n < 0) {
            cout << "Error: n debe ser un número entero no negativo" << endl;
            return 1;
        }
        
        // Medir tiempo de generación de coeficientes
        auto inicioGeneracion = high_resolution_clock::now();
        vector<long long> coeficientes = generarCoeficientesPascal(n);
        auto finGeneracion = high_resolution_clock::now();
        auto duracionGeneracion = duration_cast<microseconds>(finGeneracion - inicioGeneracion);
        
        cout << "\n=== RESULTADOS PARA n = " << n << " ===" << endl;
        cout << "Coeficientes: ";
        for (size_t i = 0; i < coeficientes.size(); i++) {
            cout << coeficientes[i];
            if (i < coeficientes.size() - 1) cout << " ";
        }
        cout << endl;
        cout << "Polinomio: " << mostrarPolinomio(coeficientes) << endl;
        cout << "Tiempo de generación: " << duracionGeneracion.count() / 1000000.0 << " segundos" << endl;
        
        // Calcular para un valor específico de x
        double x;
        cout << "\nIngrese el valor de x para calcular el polinomio: ";
        cin >> x;
        
        auto inicioCalculo = high_resolution_clock::now();
        double resultado = calcularPolinomioPasos(coeficientes, x);
        auto finCalculo = high_resolution_clock::now();
        auto duracionCalculo = duration_cast<microseconds>(finCalculo - inicioCalculo);
        
        cout << "Tiempo de cálculo: " << duracionCalculo.count() / 1000000.0 << " segundos" << endl;
        
        // Guardar resultados en archivo
        string nombreArchivo = "resultados_polinomio_n" + to_string(n) + ".txt";
        ofstream archivo(nombreArchivo);
        
        if (archivo.is_open()) {
            archivo << "RESULTADOS PARA (x+1)^" << n << endl;
            archivo << string(50, '=') << endl;
            archivo << "Coeficientes: ";
            for (size_t i = 0; i < coeficientes.size(); i++) {
                archivo << coeficientes[i];
                if (i < coeficientes.size() - 1) archivo << " ";
            }
            archivo << endl;
            archivo << "Polinomio: " << mostrarPolinomio(coeficientes) << endl;
            archivo << "Tiempo de generación: " << fixed << setprecision(6) 
                   << duracionGeneracion.count() / 1000000.0 << " segundos" << endl;
            archivo << "Valor calculado para x=" << x << ": " << resultado << endl;
            archivo << "Tiempo de cálculo: " << fixed << setprecision(6) 
                   << duracionCalculo.count() / 1000000.0 << " segundos" << endl;
            
            archivo.close();
            cout << "\nResultados guardados en '" << nombreArchivo << "'" << endl;
        } else {
            cout << "Error: No se pudo abrir el archivo para escribir" << endl;
        }
        
        // Prueba específica para n=100
        if (n == 100) {
            cout << "\n=== PRUEBA ESPECÍFICA PARA n=100 ===" << endl;
            cout << "Número de coeficientes: " << coeficientes.size() << endl;
            
            long long maxCoef = 0;
            for (long long coef : coeficientes) {
                if (coef > maxCoef) maxCoef = coef;
            }
            cout << "Coeficiente máximo: " << maxCoef << endl;
        }
        
    } catch (const exception& e) {
        cout << "Error inesperado: " << e.what() << endl;
        return 1;
    }
    
    return 0;
}